File: src/com/androhi/androiddrawableviewer/form/DetailDisplayDialog.java
package com.androhi.androiddrawableviewer.form;

import com.androhi.androiddrawableviewer.Constants;
import com.androhi.androiddrawableviewer.model.DrawableModel;
import com.androhi.androiddrawableviewer.util.IconUtils;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.ui.DialogWrapper;
import org.jetbrains.annotations.Nullable;

import javax.swing.*;
import javax.swing.border.EmptyBorder;
import javax.swing.border.SoftBevelBorder;
import java.awt.*;
import java.util.List;

public class DetailDisplayDialog extends DialogWrapper {

    private JPanel mainPanel;
    private SpringLayout layout;

    public DetailDisplayDialog(@Nullable Project project, DrawableModel drawableModel) {
        super(project, true);

        layout = new SpringLayout();
        mainPanel.setLayout(layout);
        mainPanel.setMinimumSize(new Dimension(360, 160));

        setTitle(drawableModel.getFileName());
        createContent(drawableModel);

        init();
    }

    private void createContent(DrawableModel model) {
        if (model == null) return;

        List<String> densityList = model.getDrawableDensityList();
        String fileName = model.getFileName();
        Component oldComponent = null;

        for (String density : densityList) {
            JPanel panel = new JPanel();
            panel.setLayout(new BorderLayout());

            // Label
            JLabel densityLabel = new JLabel();
            densityLabel.setText(density);
            densityLabel.setHorizontalAlignment(JLabel.CENTER);
            densityLabel.setBorder(new EmptyBorder(0, 0, 4, 0));

            // Image
            JLabel iconLabel = new JLabel();
            String filePath = model.getResourceDirectory() + Constants.PATH_SEPARATOR +
                    Constants.DRAWABLE_PREFIX + density + Constants.PATH_SEPARATOR + fileName;
            Icon icon = IconUtils.createOriginalIcon(filePath);
            iconLabel.setIcon(icon);
            iconLabel.setBorder(new SoftBevelBorder(SoftBevelBorder.LOWERED));

            panel.add(densityLabel, BorderLayout.PAGE_START);
            panel.add(iconLabel, BorderLayout.CENTER);

            String e2 = oldComponent == null ? SpringLayout.WEST : SpringLayout.EAST;
            Component c2 = oldComponent == null ? mainPanel : oldComponent;
            layout.putConstraint(SpringLayout.NORTH, panel, 8, SpringLayout.NORTH, mainPanel);
            layout.putConstraint(SpringLayout.WEST, panel, 16, e2, c2);
            mainPanel.add(panel);

            oldComponent = panel;
        }
    }

    @Nullable
    @Override
    protected JComponent createCenterPanel() {
        return mainPanel;
    }
}

File: src/main/kotlin/com/androhi/androiddrawableviewer/form/DetailDisplayDialog.kt
package com.androhi.androiddrawableviewer.form

import com.androhi.androiddrawableviewer.Constants.Companion.DRAWABLE_RESOURCE_NAME
import com.androhi.androiddrawableviewer.Constants.Companion.MIPMAP_RESOURCE_NAME
import com.androhi.androiddrawableviewer.model.DrawableModel
import com.androhi.androiddrawableviewer.utils.IconUtils
import com.intellij.openapi.project.Project
import com.intellij.openapi.ui.DialogWrapper
import com.intellij.ui.components.JBScrollPane
import java.awt.BorderLayout
import java.awt.Dimension
import javax.swing.*
import javax.swing.ScrollPaneConstants.HORIZONTAL_SCROLLBAR_ALWAYS
import javax.swing.ScrollPaneConstants.VERTICAL_SCROLLBAR_ALWAYS
import javax.swing.border.EmptyBorder
import javax.swing.border.SoftBevelBorder

class DetailDisplayDialog(project: Project, drawableModel: DrawableModel) : DialogWrapper(project, true) {

    private var mainPanel: JPanel? = null
    private val subPanel: JPanel = JPanel()
    private var springLayout: SpringLayout? = null

    init {
        springLayout = SpringLayout()

        subPanel.run {
            layout = springLayout
        }

        isAutoAdjustable = false
        setResizable(true)
        setSize(480, 360)

        title = drawableModel.fileName

        createContent(drawableModel)
        init()
    }

    private fun createContent(model: DrawableModel) {
        //addDisplayImage(model, model.fileName, Constants.DRAWABLE_PREFIX, model.drawableDensityList)
        //addDisplayImage(model, model.fileName, Constants.MIPMAP_PREFIX, model.mipmapDensityList)
        addDisplayImage(model)
    }

    // todo: すクローラブルにする。
    private fun addDisplayImage(model: DrawableModel, fileName: String = "", densityPrefix: String = "", densityList: List<String>? = null) {
        if (densityList?.isEmpty() == true) {
            return
        }

        var oldPanel: JPanel? = null
        var panelWidth = HORIZONTAL_PADDING
        var panelHeight = 0

        mainPanel?.add(createScrollPane(), BorderLayout.CENTER)

        model.filePathList.forEach { filePath ->

            val densityLabel = JLabel().apply {
                text = getDensityName(filePath)
                horizontalAlignment = JLabel.CENTER
                border = EmptyBorder(0, 0, 4, 0)
            }

            val iconLabel = JLabel().apply {
                icon = IconUtils.createOriginalIcon(filePath)
                border = SoftBevelBorder(SoftBevelBorder.LOWERED)
            }

            val panel = JPanel().apply {
                layout = BorderLayout()
                add(densityLabel, BorderLayout.PAGE_START)
                add(iconLabel, BorderLayout.CENTER)
            }

            updateContainer(panel, oldPanel)

            panelWidth += (panel.width + HORIZONTAL_PADDING)
            if (panelHeight < panel.height) {
                panelHeight = panel.height
            }

            oldPanel = panel
        }

        setContainerSize(panelWidth, panelHeight)
    }

    private fun getDensityName(filePath: String): String {
        val regex = Regex("$DRAWABLE_RESOURCE_NAME|$MIPMAP_RESOURCE_NAME|-")
        return filePath.split("/")
                .find { it.startsWith(DRAWABLE_RESOURCE_NAME) || it.startsWith(MIPMAP_RESOURCE_NAME) }
                ?.replace(regex, "")
                ?: ""
    }

    private fun createScrollPane(): JBScrollPane =
            JBScrollPane(VERTICAL_SCROLLBAR_ALWAYS, HORIZONTAL_SCROLLBAR_ALWAYS).apply {
                setViewportView(subPanel)
            }

    private fun updateContainer(newPanel: JPanel, oldPanel: JPanel?) {
        val e2 = if (oldPanel == null) SpringLayout.WEST else SpringLayout.EAST
        val c2 = if (oldPanel == null) subPanel else oldPanel
        springLayout?.run {
            putConstraint(SpringLayout.NORTH, newPanel, VERTICAL_PADDING, SpringLayout.NORTH, subPanel)
            putConstraint(SpringLayout.WEST, newPanel, HORIZONTAL_PADDING, e2, c2)
        }
        subPanel.add(newPanel)
        subPanel.doLayout()
    }

    private fun setContainerSize(width: Int, height: Int) {
        subPanel.preferredSize = Dimension(width, height + VERTICAL_PADDING * 2)
    }

    override fun createCenterPanel(): JComponent? = mainPanel

    override fun createActions(): Array<Action> = Array(1) { DialogWrapperExitAction("OK", OK_EXIT_CODE) }

    companion object {
        private const val VERTICAL_PADDING = 8
        private const val HORIZONTAL_PADDING = 16
    }
}